flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointExecutionResult.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint;
+
+/**
+ * The result of executing a checkpoint.
+ */
+public class CheckpointExecutionResult {
+
+	private final CompletedCheckpoint completedCheckpoint;
+
+	private final CheckpointAbortReason failure;
+
+	private final Throwable failureCause;
+
+	CheckpointExecutionResult(CompletedCheckpoint success) {
+		this(null, null, success);
+	}
+
+	CheckpointExecutionResult(CheckpointAbortReason failure, Throwable failureCause) {
+		this(failure, failureCause, null);
+	}
+
+	private CheckpointExecutionResult(CheckpointAbortReason failure, Throwable failureCause, CompletedCheckpoint success) {
+		this.failure = failure;
+		this.failureCause = failureCause;
+		this.completedCheckpoint = success;
```

Comment:
we could check the precondition here that we never have checkpoint and failure `!= null` at the same time.

Reply:
It is also not clear to me if or if not this class allows to have `failure != null` but `failureCause == null` and the other way around. depending on this, the precondition for `getFailureCause()` can also be adjusted to use `isFailure()`

Reply:
Yes, we allow to have `failure != null` but `failureCause == null`.

Reply:
I have already argumented that I am not in favor of having this class at all. Just as a general idea, in that case it should also be documented or even better represented by different constructors, e.g. one that only takes a `failure` but no `cause` so that user are not forced to pass in `null` into some constructors.